test/gurka/test_chinese_postman.cc

```diff
@@ -0,0 +1,574 @@
+    lls.push_back(map.nodes.at(end_node));
+  } else if (end_node.size() == 2) {
+    auto en0 = map.nodes.at(std::string(1, end_node.at(0)));
+    auto en1 = map.nodes.at(std::string(1, end_node.at(1)));
+    auto en_lat = en0.lat() + (en1.lat() - en0.lat()) / 2.0;
+    auto en_lng = en0.lng() + (en1.lng() - en0.lng()) / 2.0;
+    auto en = PointLL(en_lng, en_lat);
+
+    lls.push_back(en);
+  }
+
+  if (lls.size() != 2) {
+    throw std::logic_error("There must be 2 elements.");
+  }
+
+  rapidjson::Document doc;
+  doc.SetObject();
+  auto& allocator = doc.GetAllocator();
+  auto chinese_polygon = get_chinese_polygon(chinese_ring, allocator);
+  auto exclude_polygons = get_excluded_polys(exclude_rings, allocator);
+  auto req = build_local_req(doc, allocator, lls, costing, chinese_polygon, exclude_polygons);
+
+  return gurka::do_action(Options::chinese_postman, map, req);
+}
+
+void test_request(const gurka::map& map,
+                  const std::string& costing,
+                  const std::string& chinese_polygon_string,
+                  const std::string& exclude_polygon_string,
+                  const std::string& start_node,
+                  const std::string& end_node,
+                  const std::vector<std::string>& expected_names) {
+  auto route =
+      request_cp(map, costing, chinese_polygon_string, exclude_polygon_string, start_node, end_node);
+  gurka::assert::raw::expect_path(route, expected_names);
+}
+
+void test_request_multi_results(const gurka::map& map,
+                                const std::string& costing,
+                                const std::string& chinese_polygon_string,
+                                const std::string& exclude_polygon_string,
+                                const std::string& start_node,
+                                const std::string& end_node,
+                                const std::vector<std::vector<std::string>>& expected_names) {
+  auto route =
+      request_cp(map, costing, chinese_polygon_string, exclude_polygon_string, start_node, end_node);
+  gurka::assert::raw::expect_path_optional(route, expected_names);
+}
+} // namespace
+
+// parameterized test class to test all costings
+class ChinesePostmanTest : public ::testing::TestWithParam<std::string> {
+protected:
+  static gurka::map chinese_postman_map;
+  static gurka::map complex_chinese_postman_map;
+
+  //    A------B---<--C--->--G
+  //    |      |      |      |
+  //    |      |      ^      v
+  //    |      |      |      |
+  //    |      |      |      |
+  //    D------E------F--<---H
+
+  static void SetUpTestSuite() {
+    // Setup chinese_postman_map
+    const std::string ascii_map = R"(
+    p---------q-------r------s-----t
+    |    A----|--B----|--C---|--G  |
+    | i--|--j-|--|--k |  |   |  |  |
+    | |  |  | |  |  | |  |   |  |  |
+    | l--|--m-|--|--n |  |   |  |  |
+    |    |    |  |    |  |   |  |  |
+    |    D----|--E----|--F---|--H  |
+    u---------v-------w------x-----y
```

Comment:
Edge EF was found by boost.geometry 1.79.0 to be **within** the polygon `prwu` which it obviously is not.. 1.71.0 doesn't have that bug. I should make a small test case out of this soon and see if I can reproduce it stand-alone to file a bug.